### PR TITLE
feat: Use ape instead of brownie

### PR DESCRIPTION
**Pull request by bilbeyt**
Closes: #1079

### CODE REVIEW

**Comment by istankovic** [comment]
Regarding the state this is in, I like that we are generally in a pretty good shape here.
There are still some things to do yet, though. As we approach the maturity of the branch, I think we should
sequence the changes in separate commits, in order to facilitate further review.

Concretely, I suggest to separate into their own commit(s)

- all changes under `.circleci`
- `.gitignore` changes
- removal of `brownie-config.yaml`
- addition of `ape-config.yaml`
- changes to `pyproject.toml` and `poetry.lock`
- docs changes
- `beamer/tests` changes
- `scripts` changes

Finally, we need to get remove all changes to
- the agent code
- the contracts

File: pyproject.toml
Comment (istankovic) [inline pyproject.toml:76]: Why do we need this?

File: pyproject.toml
Comment (istankovic) [inline pyproject.toml:38]: It is strange that we are removing sphinx when we need it for our docs. What was the reason for this?

File: pyproject.toml
Comment (istankovic) [inline pyproject.toml:23]: This is fine for now, but before merging this PR, we need to make sure there's an ape release with all relevant fixes for us so that we can switch to it and not use whatever is on their `main` branch.

File: .circleci/workflows.yml
Comment (istankovic) [inline .circleci/workflows.yml:159]: Why did we drop the `-v` (verbose) flag? Doesn't ape have it?

File: .circleci/workflows.yml
Comment (istankovic) [inline .circleci/workflows.yml:180]: Same question here, as above.

File: .gitignore
Comment (istankovic) [inline .gitignore:22]: The `.DS_Store` entries should really be moved to your own personal `~/.gitignore` or something.

File: docs/source/contributing.rst
Comment (istankovic) [inline docs/source/contributing.rst:53]: There has to be a better way for this. :-)
Doesn't `ganache` support an `ape` mnemonic?
How hard would it be to add it if not?

File: scripts/_util.py
Comment (istankovic) [inline scripts/_util.py:2]: We're using Python 3.10 and above, so we don't neet `typing.Union`, we can just use the new union syntax `X | Y`.
See https://docs.python.org/3/library/stdtypes.html#types-union

File: beamer/chain.py
Comment (istankovic) [inline beamer/chain.py:6]: Please use the new union syntax `X | Y` instead of `typing.Union`.

File: docs/source/contributing.rst
Comment (bilbeyt) [reply docs/source/contributing.rst:53]: Created 2 PR for this: https://github.com/ApeWorX/ape/pull/1299 and https://github.com/ApeWorX/ape-ganache/pull/22

File: pyproject.toml
Comment (bilbeyt) [reply pyproject.toml:38]: i updated ape to use sphinx 6.1.3 and lock should have like that but i will add it here also

File: pyproject.toml
Comment (bilbeyt) [reply pyproject.toml:76]: if i recall correctly because of ape.reverts part for example, reverts is not a part of ape i guess also it is true for project.Contract i guess.

File: .gitignore
Comment (bilbeyt) [reply .gitignore:22]: yeah, it is coming all the time, i should check permanent solution for this

File: docs/source/contributing.rst
Comment (istankovic) [reply docs/source/contributing.rst:53]: Awesome.

BTW,  what about this:
> Doesn't ganache support an ape mnemonic?

Comment (bilbeyt) [reply docs/source/contributing.rst:53]: what does it mean ape mnemonic and ganache support?

Comment (istankovic) [reply docs/source/contributing.rst:53]: > what does it mean ape mnemonic and ganache support?

We now have brownie working with ganache out of the box, without us having to provide `--wallet.hdPath` and `--wallet.mnemonic` options to ganache. The question is, how can we continue not specifying these options to ganache and make ape work?

Comment (bilbeyt) [reply docs/source/contributing.rst:53]: brownie is using w3.eth.accounts to create accounts and ape is not doing that and if they do that means they are removing tx signing functionality of all TestAccounts and remaining would be like all accounts becoming ImpersonatedAccounts or so which is not exactly true in ape conditions as ImpresonatedAccounts we dont have any private keys but we have it with test accounts. We discussed this a little also in discord this is not going to happen imo.

File: pyproject.toml
Comment (bilbeyt) [reply pyproject.toml:24]: pinned ape-solidity as it wont change and for others i am following

File: pyproject.toml
Comment (istankovic) [reply pyproject.toml:78]: Yeah, most of the errors are of that sort:
```
beamer/tests/contracts/test_request_manager.py:197: error: Module has no attribute "reverts"  [attr-defined]
```

I think we definitely need to address this some way, instead of just getting rid of the error in tests like that.

There is also 
```
beamer/tests/contracts/test_request_manager.py:655: error: "ProviderAPI" has no attribute "web3"  [attr-defined]
```
as well as
```
beamer/tests/agent/unit/test_handlers.py:92: error: "Callable[[Optional[TxParams], Union[Literal['latest', 'earliest', 'pending', 'safe', 'finalized'], BlockNumber, Hash32, HexStr, HexBytes, int], Optional[Dict[ChecksumAddress, CallOverrideParams]], Opti
onal[bool]], Any]" has no attribute "return_value"  [attr-defined]
beamer/tests/agent/unit/test_handlers.py:100: error: "Callable[[Optional[TxParams]], HexBytes]" has no attribute "return_value"  [attr-defined]
beamer/tests/agent/unit/test_handlers.py:101: error: "ContractFunction" has no attribute "web3"  [attr-defined]
beamer/tests/agent/unit/test_handlers.py:112: error: "Callable[[Optional[TxParams], Union[Literal['latest', 'earliest', 'pending', 'safe', 'finalized'], BlockNumber, Hash32, HexStr, HexBytes, int], Optional[Dict[ChecksumAddress, CallOverrideParams]], Opt
ional[bool]], Any]" has no attribute "return_value"  [attr-defined]
beamer/tests/agent/unit/test_handlers.py:122: error: "ContractFunction" has no attribute "web3"  [attr-defined]
beamer/tests/conftest.py:28: error: Name "project.Resolver" is not defined  [name-defined]
beamer/tests/conftest.py:29: error: Name "project.FillManager" is not defined  [name-defined]
beamer/tests/conftest.py:30: error: Name "project.RequestManager" is not defined  [name-defined]
beamer/tests/conftest.py:31: error: Name "project.TestL1Messenger" is not defined  [name-defined]
beamer/tests/conftest.py:32: error: Name "project.TestL2Messenger" is not defined  [name-defined]
```

I think we need to think about those cases and handle them.

**Comment by istankovic** [comment]
Please move the `pylintrc` change out of `e47dd87999cdc2473016148dd749e8ae914af1e8` into its own commit.

File: ape-config.yaml
Comment (istankovic) [inline ape-config.yaml:6]: Can we also move the `.build` directory not to be on the toplevel, but under `contracts/`?

File: beamer/chain.py
Comment (istankovic) [inline beamer/chain.py:49]: Can we use just `BaseContract` from `web3.contract` here instead of a union of `Contract` and `AsyncContract`?
Those two anyway inherit from `BaseContract` and I don't think we need to differentiate between them anywhere. It would make typing less tedious.

**Comment by istankovic** [comment]
Could you please confirm that we are still waiting for next releases of
 - ape
 - ape-ganache
 
?

File: beamer/tests/conftest.py
Comment (istankovic) [inline beamer/tests/conftest.py:8]: Nitpick: I would have preferred to use full qualification for `chain` (`ape.chain`) to make it more explicit and avoid possible variable shadowing.

File: docs/source/contributing.rst
Comment (istankovic) [inline docs/source/contributing.rst:59]: Why change the key here?

File: docs/source/contributing.rst
Comment (istankovic) [inline docs/source/contributing.rst:109]: This looks like needless key change -- note that `0x1CEE82EEd89Bd5Be5bf2507a92a755dcF1D8e8dc` is already known and compromised -- we should not be making more compromised keys, intentionally or by accident. :sweat_smile:

File: ape-config.yaml
Comment (istankovic) [inline ape-config.yaml:30]: Here we are using `istanbul`. Some lines above we have `  evm_version: london`.
And in our docs we have instructions to use ganache with `london`. Which is the correct one?

File: beamer/events.py
Comment (istankovic) [inline beamer/events.py:3]: `Union` should no longer be needed since we have the new union typing syntax (`A | B`), but even that would not be need were we to use `BaseContract` as mentioned elsewhere.

File: ape-config.yaml
Comment (bilbeyt) [reply ape-config.yaml:30]: they are different things. brownie only supports istanbul as hardfork or constantinople

File: beamer/tests/contracts/test_fill_manager.py
Comment (istankovic) [inline beamer/tests/contracts/test_fill_manager.py:19]: I remember we talked about this also with ape people. Do we have an issue on ape side for making it easier to check for events?

File: docs/source/contributing.rst
Comment (bilbeyt) [reply docs/source/contributing.rst:109]: after merging ape-ganache and release, we dont need to change any docs so will discard this at some point. and only waiting ape-ganache release as ape released 0.6.3

File: beamer/chain.py
Comment (bilbeyt) [reply beamer/chain.py:49]: will try

File: beamer/tests/contracts/test_fill_manager.py
Comment (istankovic) [reply beamer/tests/contracts/test_fill_manager.py:19]: In cases like these it is better to use `any`:

```python
assert any(fill_manager.LpAdded.from_receipt(whitelist_tx))
```
because you are not creating a whole list and iterating through the entire sequence.

File: beamer/tests/contracts/test_fill_manager.py
Comment (istankovic) [inline beamer/tests/contracts/test_fill_manager.py:90]: No need to use a list here:
```python
    events = tuple(fill_manager.FillInvalidated.from_receipt(tx))
```

File: beamer/tests/contracts/test_fill_manager.py
Comment (bilbeyt) [reply beamer/tests/contracts/test_fill_manager.py:19]: i think there is .events now but maybe not like we are using before i may need to actually constract it somehow like assert LpAdded(....) in whitelist_tx.events

File: beamer/tests/contracts/test_request_manager.py
Comment (istankovic) [inline beamer/tests/contracts/test_request_manager.py:398]: Is `web3` a fixture now?

Comment (bilbeyt) [reply beamer/tests/contracts/test_request_manager.py:398]: it is hard to reach :D. `ape.chain.provider.web3`

File: ape-config.yaml
Comment (bilbeyt) [reply ape-config.yaml:6]: this is not possible with ape

File: beamer/chain.py
Comment (bilbeyt) [reply beamer/chain.py:49]: BaseContract has no .functions, it has .all_functions instead but when i tried it didnt work anyway. I guess i need to cast returned contract from w3.eth.contract to be a `Contract` and not AsyncContract

File: pyproject.toml
Comment (bilbeyt) [reply pyproject.toml]: pinned them all

File: beamer/chain.py
Comment (bilbeyt) [reply beamer/chain.py]: casted to Contract. should be better.

File: ape-config.yaml
Comment (bilbeyt) [reply ape-config.yaml:6]: now inside contracts

File: beamer/tests/contracts/test_fill_manager.py
Comment (bilbeyt) [reply beamer/tests/contracts/test_fill_manager.py]: events changed basically and now ok i guess

File: pyproject.toml
Comment (bilbeyt) [reply pyproject.toml]: you mean line by line going there and adding # type: ignore ?

Comment (istankovic) [reply pyproject.toml]: > you mean line by line going there and adding # type: ignore ?

For sure not. :-)
Need to find a saner way to deal with those.

File: pyproject.toml
Comment (istankovic) [inline pyproject.toml:24]: All of these three should be in dev-dependencies, not dependencies. They are not required for non-dev usage.

File: scripts/deployment/main.py
Comment (istankovic) [inline scripts/deployment/main.py:34]: Why this `try...except` block?

File: scripts/deployment/main.py
Comment (istankovic) [inline scripts/deployment/main.py:43]: Yes!! :heart:

File: beamer/tests/contracts/test_request_manager_fees.py
Comment (istankovic) [inline beamer/tests/contracts/test_request_manager_fees.py:65]: Beautiful and concise. :+1:

File: beamer/tests/contracts/test_l1_resolution.py
Comment (istankovic) [inline beamer/tests/contracts/test_l1_resolution.py:117]: [Totally unimportant in this particular case, but wanted to mention...]
Question: how is `b"0"` actually represented?

Spoiler: what does `tuple(b"0")` evaluate to?  ;-)

File: pyproject.toml
Comment (bilbeyt) [reply pyproject.toml]: moved to dev

File: scripts/deployment/main.py
Comment (bilbeyt) [reply scripts/deployment/main.py:34]: abstract contracts are coming in the built path also that they dont have any bytecode basically.

File: beamer/tests/contracts/test_l1_resolution.py
Comment (bilbeyt) [reply beamer/tests/contracts/test_l1_resolution.py:117]: looked at this and it is 48 which is in ascii zero :D

File: pyproject.toml
Comment (bilbeyt) [reply pyproject.toml]: still not sure about this, for example looked at brownie for reverts, i couldnt find any difference in regards with ape and brownie has no problem with reverts but ape does.

File: scripts/deployment/main.py
Comment (bilbeyt) [reply scripts/deployment/main.py]: gone

**Review by istankovic — APPROVED** [review]
Amazing work!!!

We can address the `attr-defined` and `name-defined` mypy issue after this gets merged.

**Pull request merged**
